### ISSUE TITLE
Question: How would rejoiner support getting grandparent context?

### ISSUE THREAD

**Issue by xunnanxu** [google/rejoiner#56]
In the js counterpart, there's a way to add arbitrary context info to a resolve context and thus you can relay the information all the way down as described in this issue: https://github.com/graphql/graphql-js/issues/1098

However in rejoiner since all fields are strictly typed to proto msgs or java native types, and there's no resolver exposed, it doesn't seem that use case is supported, unless one flattens the grandparent context onto the proto message (e.g. when schema is foo -> bar -> baz, let baz have foo's id), which is doable but doesn't seem like a good practice?

**Comment by siderakis** [comment]
It's possible to get the DataFetchingEnvironment (graphql.schema.DataFetchingEnvironment) in any rejoiner method (query/mutation/modification). This can be used to access a "context" object that is set when the query is executed.

Here the context is set:
https://github.com/google/rejoiner/blob/master/examples/src/main/java/com/google/api/graphql/examples/streaming/graphqlserver/GraphQlGrpcServer.java#L113

Here the DataFetchingEnvironment in injected in (it contains the context).

https://github.com/google/rejoiner/blob/master/examples/src/main/java/com/google/api/graphql/examples/streaming/graphqlserver/HelloWorldSchemaModule.java#L35

You can think of rejoiner methods as datafetchers.

I'm not 100% sure what you are trying to do, does this help?

**Comment by xunnanxu** [comment]
Hi @siderakis thanks for the quick reply.
I think unfortunately the answer is no.
Using https://graphql.org/learn/execution/#root-fields-resolvers as a reference, the `context` object here will apply to all resolvers where as the grandparent id I mentioned needs to be resolver specific (i.e. different values applied to different resolvers). The solution I originally referenced is achieved by enhancing the `obj` param in resolver, and that translated to rejoiner language means to dynamically add fields to proto message such that deeply nested object can get ids more than one level above through `@SchemaModification`. This does not seem achievable due to rejoiner only allows proto message to be returned not some interim custom data structure.

As an example assuming the query is like
```
allPeople(startDate, endDate) {
  company {
    id
    department {
      id
      team {
        id
        employee {
          id
        }
      }
    }
  }
}
```

And let's assume each level is 1-many. The `context` solution can pass `startDate` and `endDate` to each layer because they are all the same. But in case employee needs department id for grpc call that becomes tricky (assuming `employee` proto only has `team` id and `team` has `department` id). We cannot use the global context because different employee would belong to different department.

**Comment by xunnanxu** [comment]
Yeah but that bloats the return schema, which makes it messy (or you'd have to somehow remove the department id at team level to avoid that being returned to user).
`ProtoAndContext` might work but that defeats the strong typing purpose.
I wonder if it makes sense to allow methods to return arbitrary object that wraps a proto msg like
```
interface EnhancedContext<T extends Message> {
  T getMessage();
}
```
that way rejoiner can call that method to behave the same way today (and throws an exception if it can't).
